### PR TITLE
ProbabilisticScorer improvements

### PR DESCRIPTION
**Pull request by jkczyz**
In `ProbabilisticScorer`, the channel liquidity balance is reduced whenever a payment fails at the corresponding channel. The payment may still be retried through the channel, however, because the liquidity penalty is capped. Use `u64::max_value` instead in this situation to avoid retrying over the same path. This effectively makes `u64::max_value` the penalty for amounts exceeding the upper bound, as well.

Also, penalties tends to be dominated by fees for large payments. To avoid this, add an amount penalty to `ProbabilisticScorer` with a user configurable multiplier. The multiplier is applied for every `2^20`th of the amount weighted by the negative `log10` of the channel's success probability for the payment.

### CODE REVIEW

File: lightning/src/routing/scoring.rs [hunk 1]
```diff
@@ -666,19 +669,22 @@ impl<L: Deref<Target = u64>, T: Time, U: Deref<Target = T>> DirectedChannelLiqui
...
 		let max_liquidity_msat = self.max_liquidity_msat();
 		let min_liquidity_msat = core::cmp::min(self.min_liquidity_msat(), max_liquidity_msat);
-		if amount_msat > max_liquidity_msat {
-			max_penalty_msat
-		} else if amount_msat <= min_liquidity_msat {
+		if amount_msat <= min_liquidity_msat {
 			0
+		} else if amount_msat > max_liquidity_msat {
+			u64::max_value()
+		} else if amount_msat == max_liquidity_msat && max_liquidity_msat != self.capacity_msat {
```
Comment (TheBlueMatt) [inline lightning/src/routing/scoring.rs:678]: I mean, sure, but is it really worth constraining this vs just rounding off-by-one in the previous if?

**Review by TheBlueMatt — APPROVED** [review]
LG either way, but dunno if the extra branch changes anything material - was it just to simplify test changes?

File: lightning/src/routing/scoring.rs (hunk 1 above)
Comment (jkczyz) [reply lightning/src/routing/scoring.rs:678]: There are a couple edge cases to deal with.

The first mentioned in the comment is to allow at least one attempt with a non-`u64::max_value` penalty when attempting to pay with the entire effective capacity, which may be the max HTLC. We currently prefer that over capacity if both are known but should be doing the opposite. I have a commit to fix that, which we dropped from the original `ProbabilisticScorer` PR. But a user may also not have the capacity if they don't have access to the UTXO set via `chain::Access`.

The second is for private hops, where we use `u64::max_value` for the effective capacity. We'd have to use `saturated_add` so seems like a branch would be necessary. Or if we used `amount_msat  >= max_liquidity_msat` instead of adding, we'd still need to special case the first edge case, which gets a bit more messy... I suppose something like this may work since we can use `max_penalty_msat` to avoid weird logic to hit the log case?

```rust
if amount_msat >= max_liquidity_msat {
    if amount_msat > max_liquidity_msat {
        u64::max_value()
    } else if max_liquidity_msat != self.capacity_msat {
        u64::max_value()
    } else {
        max_penalty_msat
    }
}
```

That way the additional branching would be limited. How does that sound?

**Review by jkczyz — COMMENTED** [review]
> LG either way, but dunno if the extra branch changes anything material - was it just to simplify test changes?

See response to your comment which proposes an alternative to limit the extra branch.

File: lightning/src/routing/scoring.rs [hunk 2]
```diff
@@ -552,6 +555,12 @@ pub struct ProbabilisticScoringParameters {
 	/// When built with the `no-std` feature, time will never elapse. Therefore, the channel
 	/// liquidity knowledge will never decay except when the bounds cross.
 	pub liquidity_offset_half_life: Duration,
+
+	/// Whether an additional penalty should be calculated based on the payment amount in order to
+	/// provide higher penalties for large payments.
+	///
+	/// Default value: false
+	pub include_amount_penalty: bool,
```
Comment (TheBlueMatt) [inline lightning/src/routing/scoring.rs:563]: I'm not sure if we want to make this a switch instead of a second penalty - I think a second penalty more closely lines up with what fees are (a combo of % and absolute fees), which is the property we're going for here - ability to match penalties up to expected/reasonable fees.

Comment (jkczyz) [reply lightning/src/routing/scoring.rs]: Updated as discussed offline. Please take a look at the `amount_penalty_multiplier_msat` docs. I tried to make them a bit more human understandable. Let me know if I hit the mark or if it can be fine tuned in any way.

**Review by TheBlueMatt — COMMENTED** [review]
LGTM. I'd say feel free to squash.

**Review by TheBlueMatt — APPROVED** [review]

**Comment by TheBlueMatt** [comment]
LGTM. Feel free to squash.

**Review by TheBlueMatt — APPROVED** [review]

File: lightning/src/routing/scoring.rs [hunk 3]
```diff
@@ -517,7 +517,7 @@ pub struct ProbabilisticScorerUsingTime<G: Deref<Target = NetworkGraph>, T: Time
 
 /// Parameters for configuring [`ProbabilisticScorer`].
 ///
-/// Used to configure a base penalty and a liquidity penalty, the sum of which is the channel
+/// Used to configure base, liquidity, and amount penalties, the sum of which composes the channel
```
Comment (arik-so) [inline lightning/src/routing/scoring.rs:520]: s/composes/comprises

File: lightning/src/routing/scoring.rs [hunk 4]
```diff
@@ -531,7 +531,10 @@ pub struct ProbabilisticScoringParameters {
 	///
 	/// The penalty is based in part by the knowledge learned from prior successful and unsuccessful
```
Comment (arik-so) [inline lightning/src/routing/scoring.rs:532]: based in part _on_ the knowledge

File: lightning/src/routing/scoring.rs [hunk 5]
```diff
@@ -662,27 +697,62 @@ impl<T: Time> ChannelLiquidity<T> {
...
-		} else if amount_msat <= min_liquidity_msat {
+		if amount_msat <= min_liquidity_msat {
 			0
+		} else if amount_msat >= max_liquidity_msat {
+			if amount_msat > max_liquidity_msat {
+				u64::max_value()
+			} else if max_liquidity_msat != self.capacity_msat {
+				// Avoid using the failed channel on retry.
+				u64::max_value()
+			} else {
```
Comment (arik-so) [inline lightning/src/routing/scoring.rs:721]: how likely is this case to ever be hit, that amount_msat is precisely equal to max_liquidity_msat?

File: lightning/src/routing/scoring.rs [hunk 6]
```diff
@@ -662,27 +697,62 @@ impl<T: Time> ChannelLiquidity<T> {
...
-			// Upper bound the penalty to ensure some channel is selected.
-			penalty_msat.min(max_penalty_msat)
+			let negative_log10_times_1024 =
+				approx::negative_log10_times_1024(numerator, denominator);
+			self.combined_penalty_msat(amount_msat, negative_log10_times_1024, params)
 		}
 	}
 
+	#[inline(always)]
+	fn combined_penalty_msat(
```
Comment (arik-so) [inline lightning/src/routing/scoring.rs:737]: perhaps add a comment that this adds the liquidity and the amount penalties?

File: lightning/src/routing/scoring.rs (hunk 5 above)
Comment (TheBlueMatt) [reply lightning/src/routing/scoring.rs:721]: If we just tried the payment, its possible, the final else here pretty unlikely I'd guess. Maybe for "very round" numbers - 1m sats, 100k sats, etc it could happen.

Comment (jkczyz) [reply lightning/src/routing/scoring.rs:721]: Yeah, also if the HTLC maximum is some round amount, it may get hit. We should really use the capacity here if known, but currently we do not. I have a commit to do this that was dropped from the original `ProbabilisticScorer` PR.

**Review by arik-so — APPROVED** [review]
LGTM

**Review by TheBlueMatt — APPROVED** [review]
